src/dftbp/dftb/sccinit.F90

```diff
@@ -533,52 +542,50 @@ subroutine initQFromFile(qq, fileName, tReadAscii, orb, qBlock, qiBlock, density
+    end if
 
     if (tRho) then
       if (tKpointInfo) then
         coeffsAndShifts(:,:) = 0.0_dp
         if (tKpointInfoPresent) then
-          do jj = 1, size(coeffsAndShifts, dim=2)
-            do ii = 1, size(coeffsAndShifts, dim=1)
-              if (tReadAscii) then
-                read(file%unit, *, iostat=iErr) coeffsAndShifts(ii, jj)
-              else
-                read(file%unit, iostat=iErr) coeffsAndShifts(ii, jj)
-              end if
-            end do
-          end do
+          if (tReadAscii) then
+            read(file%unit, *, iostat=iErr) coeffsAndShifts
+          else
+            read(file%unit, iostat=iErr) coeffsAndShifts
+          end if
           call checkSupercellFoldingMatrix(coeffsAndShifts, errStatus,&
               & supercellFoldingDiagOut=supercellFoldingDiag)
-          allocate(densityMatrix%deltaRhoInCplxHS(orb%nOrb, orb%nOrb, supercellFoldingDiag(1),&
-              & supercellFoldingDiag(2), supercellFoldingDiag(3), nSpin))
+          if (hybridXcAlg == hybridXcAlgo%matrixBased) then
+            call getSuperSampling(coeffsAndShifts(:,1:3), modulo(coeffsAndShifts(:,4), 1.0_dp),&
+                & densityMatrix%kPointPrime, densityMatrix%kWeightPrime, reduceByInversion=.true.)
+            allocate(densityMatrix%deltaRhoInCplx(orb%nOrb, orb%nOrb,&
+                & size(densityMatrix%kPointPrime, dim=2) * nSpin))
+          else
+            allocate(densityMatrix%deltaRhoInCplxHS(orb%nOrb, orb%nOrb, supercellFoldingDiag(1),&
+                & supercellFoldingDiag(2), supercellFoldingDiag(3), nSpin))
+          end if
         end if
       end if
       if (tRhoPresent) then
         if (.not. tRealHS) then
           ! general k-point case
-          do ii = 1, size(densityMatrix%deltaRhoInCplxHS, dim=6)
-            do jj = 1, size(densityMatrix%deltaRhoInCplxHS, dim=5)
-              do kk = 1, size(densityMatrix%deltaRhoInCplxHS, dim=4)
-                do ll = 1, size(densityMatrix%deltaRhoInCplxHS, dim=3)
-                  do mm = 1, size(densityMatrix%deltaRhoInCplxHS, dim=2)
-                    do nn = 1, size(densityMatrix%deltaRhoInCplxHS, dim=1)
-                      if (tReadAscii) then
-                        read(file%unit, *, iostat=iErr)&
-                            & densityMatrix%deltaRhoInCplxHS(nn, mm, ll, kk, jj, ii)
-                      else
-                        read(file%unit, iostat=iErr)&
-                            & densityMatrix%deltaRhoInCplxHS(nn, mm, ll, kk, jj, ii)
-                      end if
-                      if (iErr /= 0) then
-                        write(error_string, *) 'Failure to read file for delta density matrix.'
-                        call error(error_string)
-                      end if
-                    end do
-                  end do
-                end do
-              end do
-            end do
-          end do
+          if (hybridXcAlg == hybridXcAlgo%matrixBased) then
```

Comment:
As this changes the charges file, bump the restartFormat version number

Reply:
Right, forgot to do that.